dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +272,371 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_HR_IF_MSG(E_UNEXPECTED, !m_comActivatorRegistration, "Need to call Register() before calling Unregister().");
+                    m_comActivatorRegistration.reset();
+                }
 
-                auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_HR_IF_MSG(E_UNEXPECTED, !m_singletonForegroundRegistration, "Need to call Register() before calling Unregister().");
+                }
 
-                wil::unique_cotaskmem_string processName;
-                THROW_IF_FAILED(GetCurrentProcessPath(processName));
-                LOG_IF_FAILED(notificationPlatform->UnregisterLongRunningActivator(processName.get()));
+                auto notificationsLongRunningPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                // Unregister foreground sink with the Long Running Process Singelton
+                THROW_IF_FAILED(notificationsLongRunningPlatform->UnregisterForegroundActivator(m_processName.get()));
 
-                s_protocolRegistration = false;
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    m_singletonForegroundRegistration = false;
+                }
             }
         }
         catch (...)
         {
-            PushNotificationTelemetry::ActivatorUnregisteredByApi(wil::ResultFromCaughtException(), activators);
+            PushNotificationTelemetry::ActivatorUnregisteredByApi(wil::ResultFromCaughtException());
             throw;
         }
 
-        PushNotificationTelemetry::ActivatorUnregisteredByApi(S_OK, activators);
+        PushNotificationTelemetry::ActivatorUnregisteredByApi(S_OK);
     }
 
-    void PushNotificationManager::UnregisterAllActivators()
+    void PushNotificationManager::UnregisterAll()
     {
-        THROW_HR_IF(E_NOTIMPL, !::Microsoft::Windows::PushNotifications::Feature_PushNotifications::IsEnabled());
+        bool unregisterCompleted{ false };
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            unregisterCompleted = (!m_singletonForegroundRegistration || !m_comActivatorRegistration);
+        }
+
+        if (!unregisterCompleted)
+        {
+            Unregister();
+        }
+
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_OPERATION_IN_PROGRESS), m_registering, "Register or Unregister currently in progress!");
+            m_registering = true;
+        }
+
+        auto scope_exit = wil::scope_exit(
+            [&] {
+                auto lock{ m_lock.lock_exclusive() };
+                m_registering = false;
+            });
 
         try
         {
-            auto lock{ s_activatorInfoLock.lock_exclusive() };
-            if (s_pushTriggerRegistration)
+            try
```

Comment:
Remove try catch. Looks like the implementation of close() swallows the error for E_NOT_SET